SYCL/ESIMD/api/functional/ctors/common.hpp

```diff
@@ -29,59 +26,58 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
-template <typename DataT, int NumElems, typename TestCaseT>
-auto call_simd(sycl::queue &queue, const shared_vector<DataT> &ref_data) {
-
-  shared_vector<DataT> result{NumElems, shared_allocator<DataT>{queue}};
-
-  queue.submit([&](sycl::handler &cgh) {
-    const auto ref = ref_data.data();
-    auto out = result.data();
-
-    cgh.single_task<Kernel<DataT, NumElems, TestCaseT>>(
-        [=]() SYCL_ESIMD_KERNEL {
-          sycl::ext::intel::experimental::esimd::simd<DataT, NumElems>
-              result_simd =
-                  TestCaseT::template call_simd_ctor<DataT, NumElems>(ref);
-          result_simd.copy_to(out);
-        });
-  });
-  return result;
-}
+template <typename DataT, int NumElems, typename ContextT>
+class TestDescription : public ITestDescription {
+protected:
+  const std::string m_data_type;
+  const DataT m_retrieved_val;
+  const DataT m_expected_val;
+  const size_t m_index;
+
+public:
+  TestDescription(size_t index, DataT retrieved_val, DataT expected_val,
+                  const std::string &data_type)
+      : m_data_type(data_type), m_retrieved_val(retrieved_val),
+        m_expected_val(expected_val), m_index(index) {}
+
+  std::string to_string() const override {
+    // TODO: Make strings for fp values more short during failure output, may be
+    // by using hex representation
+    std::string log_msg("Failed for simd<");
+
+    log_msg += m_data_type + ", " + std::to_string(NumElems) + ">";
+    log_msg += ", with context: " + ContextT::get_description();
+    log_msg += ", retrieved: " + std::to_string(m_retrieved_val);
+    log_msg += ", expected: " + std::to_string(m_expected_val);
+    log_msg += ", at index: " + std::to_string(m_index);
+
+    return log_msg;
+  }
+};
 
 // The main test routine.
 // Using functor class to be able to iterate over the pre-defined data types.
-template <typename DataT, int NumElems, typename TestCaseT> struct test {
+template <typename DataT, int NumElems, typename TestCaseT, typename VerifierT>
+struct run_test {
   bool operator()(sycl::queue &queue, const std::string &data_type) {
-    bool passed{true};
-
-    std::vector<DataT> generated_data{generate_ref_data<DataT, NumElems>()};
-    shared_vector<DataT> ref_data{generated_data.begin(), generated_data.end(),
-                                  shared_allocator<DataT>{queue}};
-
-    const auto result_data =
-        call_simd<DataT, NumElems, TestCaseT>(queue, ref_data);
-
-    for (size_t it = 0; it < ref_data.size(); it++) {
-      if (!are_bitwise_equal(ref_data[it], result_data[it])) {
-        passed = false;
-        log::fail<NumElems>(
-            "Simd by " + TestCaseT::get_description() +
-                " failed, retrieved: " + std::to_string(result_data[it]) +
-                ", expected: " + std::to_string(ref_data[it]),
-            data_type);
+    bool passed = true;
+    std::vector<DataT> ref_data{generate_ref_data<DataT, NumElems>()};
```

Comment:
```suggestion
    const std::vector<DataT> ref_data = generate_ref_data<DataT, NumElems>();
```

Reply:
I've applied this comment in this commit: a6fac3bb1b02498714bd0edbdd9648053e8a27f7